janus_server/src/metrics.rs

```diff
@@ -0,0 +1,188 @@
+}
+
+/// Selection of an exporter for OpenTelemetry metrics.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum MetricsExporterConfiguration {
+    #[serde(rename = "prometheus")]
+    Prometheus {
+        host: Option<String>,
+        port: Option<u16>,
+    },
+    #[serde(rename = "otlp")]
+    Otlp(OtlpExporterConfiguration),
+}
+
+/// Configuration options specific to the OpenTelemetry OTLP metrics exporter.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct OtlpExporterConfiguration {
+    /// gRPC endpoint for OTLP exporter.
+    pub endpoint: String,
+    /// Additional metadata/HTTP headers to be sent with OTLP requests.
+    #[serde(default)]
+    pub metadata: HashMap<String, String>,
+}
+
+/// Choice of OpenTelemetry metrics exporter implementation.
+pub enum MetricsExporterHandle {
+    #[cfg(feature = "prometheus")]
+    Prometheus(tokio::task::JoinHandle<()>),
+    #[cfg(feature = "otlp")]
+    Otlp(opentelemetry::sdk::metrics::controllers::PushController),
+    Noop,
+}
+
+#[derive(Debug)]
+struct CustomAggregatorSelector;
+
+impl AggregatorSelector for CustomAggregatorSelector {
+    fn aggregator_for(&self, descriptor: &Descriptor) -> Option<Arc<dyn Aggregator + Send + Sync>> {
+        match descriptor.instrument_kind() {
+            opentelemetry::metrics::InstrumentKind::ValueRecorder => {
+                // The following boundaries are copied from the Ruby and Go Prometheus clients.
+                // Buckets could be specialized per-metric by matching on `descriptor.name()`.
+                let boundaries = &[
+                    0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0,
+                ][..];
+                Some(Arc::new(
+                    opentelemetry::sdk::metrics::aggregators::histogram(descriptor, boundaries),
+                ))
+            }
+            opentelemetry::metrics::InstrumentKind::ValueObserver => Some(Arc::new(
+                opentelemetry::sdk::metrics::aggregators::last_value(),
+            )),
+            _ => Some(Arc::new(opentelemetry::sdk::metrics::aggregators::sum())),
+        }
+    }
+}
+
+/// Install a metrics provider and exporter, per the given configuration. The OpenTelemetry global
+/// API can be used to create and update meters, and they will be sent through this exporter. The
+/// returned handle should not be dropped until the application shuts down.
+pub fn install_metrics_exporter(
+    config: &MetricsConfiguration,
+) -> Result<MetricsExporterHandle, Error> {
+    match &config.exporter {
+        #[cfg(feature = "prometheus")]
+        Some(MetricsExporterConfiguration::Prometheus {
+            host: config_exporter_host,
+            port: config_exporter_port,
+        }) => {
+            use http::StatusCode;
+            use hyper::Response;
+            use prometheus::{Encoder, TextEncoder};
+            use std::net::SocketAddr;
+            use warp::{Filter, Reply};
```

Comment:
I think moving these `use` statements (as well as the ones in other arms of this `match`) would declutter this method and make it easier to read. If you want to make the `use` conditional you can still do

```
#[cfg(feature="something")]
use {
    http::{...},
    hyper::{...},
    etc.
};
```